File: QytetetRuby/lib/especulador.rb
#encoding: utf-8
require_relative "jugador.rb"

module ModeloQytetet
  class Especulador < Jugador
    attr_accessor :fianza
    
    def initialize(jugador, fianza)
       super(jugador.nombre,jugador.saldo,jugador.encarcelado, jugador.propiedades, jugador.cartaLibertad, jugador.casillaActual)
       @fianza = fianza
    end
    
    def pagarImpuesto
      modificarSaldo(-((@casillaActual.coste)/2))
    end
    
    def convertirme(fianza)
      return self
    end
    
    def deboIrACarcel
      deboIr = false
      
      if (super == true) && (pagarFianza == false)
        deboIr = true
      end
      
      return deboIr
    end
    
    def pagarFianza
      pagada = false
      
      if tengoSaldo(@fianza)
        modificarSaldo(-@fianza)
        puts "\nTe has librado de ir a la carcel\n"
        pagada = true
      end
      
      return pagada
    end
    
    def puedoEdificarCasa(titulo) 
      puedo_edificar = false
      
      if titulo.numCasas < 8
        puedo_edificar = true
      end
      
      return puedo_edificar
    end
    
    def puedoEdificarHotel(titulo)
      puedo_edificar = false
      
      if titulo.numHoteles < 8 && titulo.numCasas >= 4
        puedo_edificar = true
      end
      
      return puedo_edificar
    end
    
    def to_s
      "Especulador{"+super.to_s + "fianza=" + @fianza.to_s + "}"
    end
    
    protected :pagarImpuesto, :initialize,  :deboIrACarcel,
               :puedoEdificarCasa, :puedoEdificarHotel
    private :pagarFianza
  end
end

File: QytetetJava/src/modeloqytetet/Especulador.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package modeloqytetet;

/**
 *
 * @author Juan Carlos Hermoso Quesada
 */
public class Especulador extends Jugador{
    private int fianza;
    
    public Especulador(int fianza, Jugador jugador) {
        super(jugador);
        this.fianza = fianza;
    }
    
    @Override
    protected void pagarImpuesto() {
        modificarSaldo(-(super.getCasillaActual().getCoste())/2);
    }
        
    @Override
    protected Especulador convertirme(int fianza) {
        return this;
    }
    
    @Override
    protected boolean deboIrACarcel() {
        boolean deboIrACarcel = false;
        
        if (super.deboIrACarcel() == true && pagarFianza() == false) {
            deboIrACarcel = true;
        }
        
        return deboIrACarcel;
    }
    
    private boolean pagarFianza() {
        boolean pagada = false;
        
        if (tengoSaldo(fianza)) {
            System.out.println("\nTe has librado de ir a la cárcel\n");
            modificarSaldo(-fianza);
            pagada = true;
        }
        
        return pagada;
    }
    
    @Override
    protected boolean puedoEdificarCasa(TituloPropiedad titulo) {
        boolean puedo_edificar = false;
        
        if (titulo.getNumCasas() < 8) {
            puedo_edificar = true;
        }
        
        return puedo_edificar;
    }
    
    @Override
    protected boolean puedoEdificarHotel(TituloPropiedad titulo) {
        boolean puedo_edificar = false;
        
        if (titulo.getNumHoteles() < 8 && titulo.getNumCasas() >= 4) {
            puedo_edificar = true;
        }
        
        return puedo_edificar;
    }

    @Override
    public String toString() {
        return "Especulador "+super.toString() + "\nFianza= " + fianza;
    }
}
